src/generic/numberparsing.h

```diff
@@ -1,80 +1,1145 @@
+  // this will be non-zero because the most significant bit of i is
+  // 1.
+  value128 product = full_multiplication(i,factor_mantissa);
+  uint64_t lower = product.low;
+  uint64_t upper = product.high;
+
+  // We know that upper has at most one leading zero because
+  // both i and  factor_mantissa have a leading one. This means
+  // that the result is at least as large as ((1<<63)*(1<<63))/(1<<64).
+
+  // As long as the first 9 bits of "upper" are not "1", then we
+  // know that we have an exact computed value for the leading
+  // 55 bits because any imprecision would play out as a +1, in
+  // the worst case.
+  if (unlikely((upper & 0x1FF) == 0x1FF) && (lower + i < lower)) {
+    uint64_t factor_mantissa_low =
+        mantissa_128[power - FASTFLOAT_SMALLEST_POWER];
+    // next, we compute the 64-bit x 128-bit multiplication, getting a 192-bit
+    // result (three 64-bit values)
+    product = full_multiplication(i,factor_mantissa_low);
+    uint64_t product_low = product.low;
+    uint64_t product_middle2 = product.high;
+    uint64_t product_middle1 = lower;
+    uint64_t product_high = upper;
+    uint64_t product_middle = product_middle1 + product_middle2;
+    if (product_middle < product_middle1) {
+      product_high++; // overflow carry
+    }
+    // we want to check whether mantissa *i + i would affect our result
+    if (((product_middle + 1 == 0) && ((product_high & 0x1FF) == 0x1FF) &&
+         (product_low + i < product_low))) { // let us be prudent and bail out.
+      *success = false;
+      return 0;
+    }
+    upper = product_high;
+    lower = product_middle;
+  }
+  // The final mantissa should be 53 bits with a leading 1.
+  // We shift it so that it occupies 54 bits with a leading 1.
+  ///////
+  uint64_t upperbit = upper >> 63;
+  uint64_t mantissa = upper >> (upperbit + 9);
+  lz += 1 ^ upperbit;
+
+  // Here we have mantissa < (1<<54).
+
+  // We have to round to even. The "to even" part
+  // is only a problem when we are right in between two floats
+  // which we guard against.
+  // If we have lots of trailing zeros so we may fall right between two
+  // floating-point values.
+  if (unlikely((lower == 0) && ((upper & 0x1FF) == 0) &&
+               ((mantissa & 1) == 1))) {
+    // It can be triggered with 1e23.
+    *success = false;
+    return 0;
+  }
+  mantissa += mantissa & 1;
+  mantissa >>= 1;
+  // Here we have mantissa < (1<<53), unless there was an overflow
+  if (mantissa >= (1ULL << 53)) {
+    //////////
+    // This will happen when parsing values such as 7.2057594037927933e+16
+    ////////
+    mantissa = (1ULL << 52);
+    lz--; // undo previous addition
+  }
+  mantissa &= ~(1ULL << 52);
+  uint64_t real_exponent = c.exp - lz;
+  // we have to check that real_exponent is in range, otherwise we bail out
+  if (unlikely((real_exponent < 1) || (real_exponent > 2046))) {
+    *success = false;
+    return 0;
+  }
+  mantissa |= real_e
```

Comment:
what about super large values, would strtod return infinity for those? does the end behavior differ from other json parsers?

Reply:
> what about super large values, would strtod return infinity for those? does the end behavior differ from other json parsers?

Values like infinity or NaN are not allowed in the JSON specification. If you eat up a large value and you map it to "infinity", you will no longer be able to serialize back a standard-compliant JSON. In any case, there is no realistic application where you might need values so large than they can't fit in binary64. If this ever happens, you know that you have a bug. 


RapidJSON (the most standard compliant of the "other parsers") will reject a document with a float that does not fit in binary64. JSON for Modern C++ (nlohmann/json) will flat out throw an exception.

Reply:
@michaeleisel There are 10^80 electrons in the universe. If you find a value larger than 10^308 anywhere in software, it is surely a bug.

I will document the issue in a comment.